federationsender/internal/api.go

```diff
@@ -38,3 +44,98 @@ func NewFederationSenderInternalAPI(
 		queues:     queues,
 	}
 }
+
+func (a *FederationSenderInternalAPI) isBlacklistedOrBackingOff(s gomatrixserverlib.ServerName) (*statistics.ServerStatistics, error) {
+	stats := a.statistics.ForServer(s)
+	if stats.Blacklisted() {
+		return stats, &api.FederationClientError{
+			Blacklisted: true,
+		}
+	}
+	// Call BackoffIfRequired with a closed channel to make it return immediately.
+	// It will return the duration to backoff for.
+	var duration time.Duration
+	interrupt := make(chan bool)
+	close(interrupt)
+	var bo atomic.Bool
+	duration, _ = stats.BackoffIfRequired(bo, interrupt)
+	if duration > 0 {
+		return stats, &api.FederationClientError{
+			RetryAfter: duration,
+		}
+	}
+
+	return stats, nil
+}
+
+func failBlacklistableError(err error, stats *statistics.ServerStatistics) {
+	if err == nil {
+		return
+	}
+	mxerr, ok := err.(gomatrix.HTTPError)
+	if !ok {
+		stats.Failure()
+		return
+	}
+	if mxerr.Code == 500 {
```

Comment:
This should probably check for any error in the 500s, rather than just 500 itself.

Reply:
Not sure on this one, a 502 Bad Gateway because of apache/nginx fails feels wrong? 503/504 though seems ok?

Reply:
I would think that we should still backoff for 502, as if you are using an nginx reverse proxy, Cloudflare or similar (as I am) then a bad gateway error shouldn't result in us giving up.